File: dailyQuestion/2020-05/05-30/python/solution_mono_stack_const.py
# 单调栈 + 常数优化
def largestRectangleArea(heights: [int]) -> int:
    n = len(heights)
    left = [0] * n
    right = [n] * n
    mono_stack = []
    for i in range(n):
        while mono_stack and heights[mono_stack[-1]] >= heights[i]:
            right[mono_stack[-1]] = i
            mono_stack.pop()
        left[i] = mono_stack[-1] if mono_stack else -1
        mono_stack.append(i)
    result = max((right[i] - left[i] - 1) * heights[i] for i in range(n)) if heights else 0
    return result


if __name__ == "__main__":
    heights = [2, 1, 5, 6, 2, 3]
    result = largestRectangleArea(heights)
    print(result)

File: dailyQuestion/2020-05/05-30/golang/solution_mono_stack_const.go
package main

import (
	"fmt"
)

func main() {
	heights := []int{6, 7, 5, 2, 4, 5, 9, 3}
	result := largestRectangleArea(heights)
	fmt.Println(result)

}
func largestRectangleArea(heights []int) int {
	n := len(heights)
	if n == 0 {
		return 0
	}
	left := make([]int, n)
	right := make([]int, n)
	for i := 0; i < n; i++ {
		right[i] = n
	}
	mono_stack := []int{}
	for i := 0; i < n; i++ {
		for len(mono_stack) != 0 && heights[mono_stack[len(mono_stack)-1]] >= heights[i] {
			right[mono_stack[len(mono_stack)-1]] = i
			mono_stack = mono_stack[:len(mono_stack)-1]
		}
		if len(mono_stack) > 0 {
			left[i] = mono_stack[len(mono_stack)-1]
		} else {
			left[i] = -1
		}
		mono_stack = append(mono_stack, i)
	}
	result := 0
	for i := 0; i < n; i++ {
		result = max(result, (right[i]-left[i]-1)*heights[i])
	}
	return result
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
